### PR TITLE
AGAS Refactoring+1753 Cache mods

### LINKED ISSUE #279: Refactor addressing_service into a base class and two derived classes
**[reported by blelbach]** *[Trac time Sat Dec 10 21:31:40 2011]* addressing_service should be implemented with a class hierarchy (one derived class for hosted, one derived class for bootstrap) to simplify the code.

### PR DESCRIPTION
**Pull request by parsa**
This fixes #279

### CODE REVIEW

File: hpx/runtime/agas/addressing_service.hpp [hunk 1]
```diff
@@ -12,6 +14,7 @@
 #include <hpx/config.hpp>
 
 #include <hpx/exception.hpp>
+#include <hpx/hpx_fwd.hpp>
```
Comment (sithhell) [inline hpx/runtime/agas/addressing_service.hpp]: Could this be avoided? I would rather prefer for new code to remove hpx_fwd.hpp.

Comment (hkaiser) [inline hpx/runtime/agas/addressing_service.hpp]: Yes, it's a merge artifact, I'll remove it.

File: hpx/runtime/agas/detail/agas_service_client.hpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+#include <hpx/runtime/agas/response.hpp>
+#include <hpx/lcos/future.hpp>
+
+#include <vector>
+#include <string>
+
+namespace hpx { namespace agas { namespace detail
+{
+    struct bootstrap_data_type;
+    struct hosted_data_type;
```
Comment (sithhell) [inline hpx/runtime/agas/detail/agas_service_client.hpp]: Why are those forward declarations needed here?

Comment (hkaiser) [inline hpx/runtime/agas/detail/agas_service_client.hpp]: They are not needed anymore, indeed. Good catch!

**Pull request merged**